plugins/out_es/es_conf.c

```diff
@@ -17,521 +17,899 @@
+            return -1;
         }
-        else {
-            flb_plg_error(ins, "wrong Write_Operation (should be one of index, create, update, upsert)");
-            flb_es_conf_destroy(ctx);
-            return NULL;
+    }
+
+    tmp = flb_upstream_node_get_property(FLB_ES_CONFIG_PROPERTY_ID_KEY, node);
+    if (tmp) {
+        ec->id_key = flb_sds_create(tmp);
+        if (ec->id_key == NULL) {
+            return -1;
         }
-        if (strcasecmp(ctx->es_action, FLB_ES_WRITE_OP_UPDATE) == 0
-            && !ctx->ra_id_key && ctx->generate_id == FLB_FALSE) {
-            flb_plg_error(ins, "Id_Key or Generate_Id must be set when Write_Operation update or upsert");
-            flb_es_conf_destroy(ctx);
-            return NULL;
+        ec->own_id_key = FLB_TRUE;
+        ret = config_set_ra_id_key(ec->id_key, ec, ctx);
+        if (ret != 0) {
+            return -1;
         }
     }
 
-    if (ctx->logstash_prefix_key) {
-        if (ctx->logstash_prefix_key[0] != '$') {
-            len = flb_sds_len(ctx->logstash_prefix_key);
-            buf = flb_malloc(len + 2);
-            if (!buf) {
-                flb_errno();
-                flb_es_conf_destroy(ctx);
-                return NULL;
-            }
-            buf[0] = '$';
-            memcpy(buf + 1, ctx->logstash_prefix_key, len);
-            buf[len + 1] = '\0';
-
-            ctx->ra_prefix_key = flb_ra_create(buf, FLB_TRUE);
-            flb_free(buf);
+    tmp = flb_upstream_node_get_property(FLB_ES_CONFIG_PROPERTY_WRITE_OPERATION, node);
+    if (tmp) {
+        ec->write_operation = flb_sds_create(tmp);
+        if (ec->write_operation == NULL) {
+            return -1;
         }
-        else {
-            ctx->ra_prefix_key = flb_ra_create(ctx->logstash_prefix_key, FLB_TRUE);
+        ec->own_write_operation = FLB_TRUE;
+        ret = config_set_es_action(ec->write_operation, ec->ra_id_key, ec->generate_id,
+                                   ec, ctx);
+
+        if (ret != 0) {
+            return -1;
         }
+    }
 
-        if (!ctx->ra_prefix_key) {
-            flb_plg_error(ins, "invalid logstash_prefix_key pattern '%s'", tmp);
-            flb_es_conf_destroy(ctx);
-            return NULL;
+#ifdef FLB_HAVE_AWS
+    if ((base->has_aws_auth != ec->has_aws_auth)
+        || (base->has_aws_auth == FLB_TRUE
+            && ec->has_aws_auth == FLB_TRUE
+            && aws_provider_node == FLB_TRUE)) {
+        ret = flb_es_conf_set_aws_provider(aws_external_id, aws_role_arn, ec, ctx,
+                                           config);
+        if (ret != 0) {
+            flb_plg_error(ctx->ins, "cannot configure AWS authentication");
+            return -1;
         }
     }
+#endif
+
+    return 0;
+}
+
+static void elasticsearch_config_destroy(struct flb_elasticsearch_config *ec)
+{
+    if (ec->own_tag_key == FLB_TRUE) {
+        flb_sds_destroy(ec->tag_key);
+    }
+
+    if (ec->ra_id_key && ec->own_ra_id_key == FLB_TRUE) {
```

Comment:
We can distinguish whether ec->ra_id_key with if condition when initialized with NULL on initialize.

Reply:
Answered in [another comment](https://github.com/fluent/fluent-bit/pull/7608#issuecomment-2526291616). `own_xxx` fields are used to deallocate data shared b/w plugin configuration and upstream node configuration just once.